x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/Join.java

```diff
@@ -68,47 +73,35 @@ public Join replaceChildren(LogicalPlan left, LogicalPlan right) {
         }
         return lazyOutput;
     }
 
-    private List<Attribute> computeOutput() {
-        List<Attribute> right = makeReference(right().output());
+    /**
+     * Merge output fields.
+     * Currently only implemented for LEFT JOINs; the rightOutput shadows the leftOutput, except for any attributes that
+     * occur in the join's matchFields.
+     */
+    public static List<Attribute> computeOutput(List<Attribute> leftOutput, List<Attribute> rightOutput, JoinConfig config) {
+        AttributeSet matchFieldSet = new AttributeSet(config.matchFields());
+        Set<String> matchFieldNames = new HashSet<>(Expressions.names(config.matchFields()));
         return switch (config.type()) {
-            case LEFT -> // right side becomes nullable
-                mergeOutput(left().output(), makeNullable(right), config.matchFields());
-            case RIGHT -> // left side becomes nullable
-                mergeOutput(makeNullable(left().output()), right, config.matchFields());
-            case FULL -> // both sides become nullable
-                mergeOutput(makeNullable(left().output()), makeNullable(right), config.matchFields());
-            default -> // neither side becomes nullable
-                mergeOutput(left().output(), right, config.matchFields());
+            case LEFT -> {
+                // Right side becomes nullable.
+                List<Attribute> fieldsAddedFromRight = removeCollisionsWithMatchFields(rightOutput, matchFieldSet, matchFieldNames);
+                yield mergeOutputAttributes(makeNullable(makeReference(fieldsAddedFromRight)), leftOutput);
+            }
+            default -> throw new UnsupportedOperationException("Other JOINs than LEFT not supported");
         };
     }
 
-    /**
-     * Merge output fields, left hand side wins in name conflicts <strong>except</strong>
-     * for fields defined in {@link JoinConfig#matchFields()}.
-     */
-    public static List<Attribute> mergeOutput(
-        List<? extends Attribute> lhs,
-        List<? extends Attribute> rhs,
-        List<NamedExpression> matchFields
+    private static List<Attribute> removeCollisionsWithMatchFields(
+        List<Attribute> attributes,
+        AttributeSet matchFields,
+        Set<String> matchFieldNames
     ) {
-        List<Attribute> results = new ArrayList<>(lhs.size() + rhs.size());
-
-        for (Attribute a : lhs) {
-            if (rhs.contains(a) == false || matchFields.stream().anyMatch(m -> m.name().equals(a.name()))) {
-                results.add(a);
-            }
-        }
-        for (Attribute a : rhs) {
-            if (false == matchFields.stream().anyMatch(m -> m.name().equals(a.name()))) {
-                results.add(a);
-            }
-        }
-        return results;
+        return attributes.stream().filter(attr -> matchFields.contains(attr) || matchFieldNames.contains(attr.name()) == false).toList();
```

Comment:
We tend to avoid using streams, except tests, as it's an unnecessary runtime complication.